working/333 - shared memory multithreading/proposal.md

```diff
@@ -0,0 +1,591 @@
+  await Isolate.runShared(() {
+    print(global);  // IsolationError: Can't access 'global' when running in shared isolate
+    global = 24;  // IsolationError: Can't access 'global' when running in shared isolate
+
+    print(sharedGlobal);  // => 42
+    sharedGlobal = 24;
+  });
+  print(global);  // => 42
+  print(sharedGlobal);  // => 24
+}
+```
+
+### Why not compile time isolation?
+
+It is tempting to try introducing a compile time separation between functions which only access `shared` state and functions which can access isolated state.
+
+One obvious approach is to introduce a modifier (e.g. `shared`) which can be applied to function declarations and impose a number of restrictions that `shared` functions have to satisfy. These restrictions should guarantee that `shared` functions can only access `shared` state.
+
+```dart
+shared void foo() {
+  // ...
+}
+```
+
+* You can only pass subtypes of `Shareable` to `shared` methods and you can only get a `Shareable` result back.
+
+  * Consequently an instance method can only be marked  `shared` if it is declared in a `Shareable` class.
+
+* You can't override non-`shared` method with `shared` method.
+* Within a `shared` function
+  * If `f(...)` is a static function invocation then `f` must be a `shared` function.
+  * If `o.m(...)` is an instance method invocation, then  `o` must be a subtype of `Shareable` and `m` must be `shared` method.
+  * If `C(...)` is a constructor invocation then `C` must be a sharable class.
+  * If `g` is a reference to a global variable then `g` must be `shared`.
+
+This approach seems promising on the surface, but we quickly hit issues:
+
+* It's unclear how to treat `Object` members like `toString`, `operator ==` and `get hashCode`. These can't be marked as `shared` but should be accessible to both `shared` and non-`shared` code.
+* It's unclear how to treat function expression invocations:
+  * Function types don't encode necessary separation between `shared` and non-`shared` functions.
+  * Methods like `List<T>.forEach` pose challenge because they should be usable in both `shared` and non-`shared` contexts.
+
+This makes us think that language changes required to achieve sound compile time delineation between `shared` and isolate worlds are too complicated to be worth it.
+
+###  Upgrading `dart:ffi`
+
+Introduction of _shared isolate_ allows to finally address the problem of native code invoking Dart callbacks from arbitrary threads. [`NativeCallable`](https://api.dart.dev/dev/3.3.0-246.0.dev/dart-ffi/NativeCallable-class.html) can be extended with the corresponding constructor:
+
+```dart
+class NativeCallable<T extends Function> {
+  /// Constructs a [NativeCallable] that can be invoked from any thread.
+  ///
+  /// When the native code invokes the function [nativeFunction], the corresponding
+  /// [callback] will be synchronously executed on the same thread within a
+  /// shared isolate corresponding to the current isolate group.
```

Comment:
Should native code be aware of isolate groups, should it worry about what isolate group is current?

Reply:
No. We have FfiCallbackMetadata which maps the entry-point of callbacks to the isolate that created the callback.

Reply:
Added test to clarify this.